grammar.js

```diff
@@ -1339,6 +1452,129 @@ module.exports = grammar({
+            $.keyword_all,
+            field("configuration_parameter", $.identifier)
+          ),
+        ),
+        seq(
+          $.keyword_set,
+          choice(
+            seq($.keyword_tablespace, $.identifier),
+              $.set_configuration,
+            ),
+          ),
+        ),
+      ),
+
+    alter_role: $ => seq(
+      $.keyword_alter,
+      choice(
+        $.keyword_role,
+        $.keyword_group,
+        $.keyword_user,
+      ),
+      choice($.identifier, $.keyword_all),
+      choice(
+        $.rename_object,
+        seq(optional($.keyword_with),repeat($._role_options)),
+        seq(
+          optional(seq($.keyword_in, $.keyword_database, $.identifier)),
+          choice(
+            seq(
+              $.keyword_set,
+              $.set_configuration,
+            ),
+            seq(
+              $.keyword_reset,
+              choice(
+                $.keyword_all,
+                field("option", $.identifier),
+              )),
+          ),
+        )
+      ),
+    ),
+
+    set_configuration: $ => seq(
+      field("option", $.identifier),
+      choice(
+        seq($.keyword_from, $.keyword_current),
+        seq(
+          choice($.keyword_to, "="),
+          choice(
+            field("parameter", $.identifier),
+            $.literal,
+            $.keyword_default
+          )
+        )
+      ),
+    ),
+
+    alter_index: $ => seq(
+      $.keyword_alter,
+      $.keyword_index,
+      optional($._if_exists),
+      $.identifier,
+      choice(
+        $.rename_object,
+        seq(
+          $.keyword_alter,
+          optional($.keyword_column),
+          alias($._natural_number, $.literal),
+          $.keyword_set,
+          $.keyword_statistics,
+          alias($._natural_number, $.literal),
+        ),
+        seq($.keyword_reset, paren_list($.identifier)),
+        seq(
+          $.keyword_set,
+          choice(
+            seq($.keyword_tablespace, $.identifier),
+            paren_list(seq($.identifier, '=', field("value", $.literal)))
+          ),
+        ),
+      ),
+    ),
+
+    alter_sequence: $ => seq(
+      $.keyword_alter,
+      $.keyword_sequence,
+      optional($._if_exists),
+      $.object_reference,
+      choice(
+        seq(
+          optional(seq($.keyword_as, $._type)),
+          optional(seq($.keyword_increment, optional($.keyword_by), $.literal)),
+          optional(seq($.keyword_minvalue, choice($.literal, seq($.keyword_no, $.keyword_minvalue)))),
+          optional(seq($.keyword_maxvalue, choice($.literal, seq($.keyword_no, $.keyword_maxvalue)))),
+          optional(seq($.keyword_start, optional($.keyword_with), field("start", alias($._integer, $.literal)))),
+          optional(seq($.keyword_restart, optional($.keyword_with), field("restart", alias($._integer, $.literal)))),
+          optional(seq($.keyword_cache, field("cache", alias($._integer, $.literal)), optional(seq(optional($.keyword_no), $.keyword_cycle)))),
```

Comment:
cache and cycle are separate (`alter sequence x no cycle` is allowed), otherwise this is the correct version of this block -- the "create" version still has both cache and cycle baked into `start`

Reply:
Ah it o weird that in the documentation, they put these optional sections in one line. That is a bit confusing.

I have fixed that.

Reply:
the `(no) cycle` path is having a really outsized effect on build time and state count -- just commenting those out in create/alter sequence brings us down from 8313 states to 6143, and build times back to 50s on my machine. But all those optionals do a lot: if you comment out the `as type` in `create sequence` as well we're down to 5787 states, comment it out in `alter sequence` instead 6021 (I expected `alter` to be heavier since it's more deeply nested, weird!).

But it seems Postgres isn't too particular about the exact order: `alter sequence s no cycle restart with 1` is accepted. So we can do this instead:

```javascript
    create_sequence: $ => seq(
      $.keyword_create,
      optional(
        choice(
          choice($.keyword_temporary, $.keyword_temp),
          $.keyword_unlogged,
        )
      ),
      $.keyword_sequence,
      optional($._if_not_exists),
      $.object_reference,
      repeat1(
        choice(
          seq($.keyword_as, $._type),
          seq($.keyword_increment, optional($.keyword_by), field("increment", alias($._integer, $.literal))),
          seq($.keyword_minvalue, choice($.literal, seq($.keyword_no, $.keyword_minvalue))),
          seq($.keyword_maxvalue, choice($.literal, seq($.keyword_no, $.keyword_maxvalue))),
          seq($.keyword_start, optional($.keyword_with), field("start", alias($._integer, $.literal))),
          seq($.keyword_cache, field("cache", alias($._integer, $.literal))),
          seq(optional($.keyword_no), $.keyword_cycle),
          seq($.keyword_owned, $.keyword_by, choice($.keyword_none, $.object_reference)),
        ),
      ),
    ),

    alter_sequence: $ => seq(
      $.keyword_alter,
      $.keyword_sequence,
      optional($._if_exists),
      $.object_reference,
      choice(
        repeat1(
          choice(
            seq($.keyword_as, $._type),
            seq($.keyword_increment, optional($.keyword_by), $.literal),
            seq($.keyword_minvalue, choice($.literal, seq($.keyword_no, $.keyword_minvalue))),
            seq($.keyword_maxvalue, choice($.literal, seq($.keyword_no, $.keyword_maxvalue))),
            seq($.keyword_start, optional($.keyword_with), field("start", alias($._integer, $.literal))),
            seq($.keyword_restart, optional($.keyword_with), field("restart", alias($._integer, $.literal))),
            seq($.keyword_cache, field("cache", alias($._integer, $.literal))),
            seq(optional($.keyword_no), $.keyword_cycle),
            seq($.keyword_owned, $.keyword_by, choice($.keyword_none, $.object_reference)),
          ),
        ),
        $.rename_object,
        $.change_ownership,
        seq(
          $.keyword_set,
          choice(
            choice($.keyword_logged, $.keyword_unlogged),
            seq($.keyword_schema, $.identifier)
          ),
        ),
      ),
    ),
```

and wind up with 4275 states and 36s build time

Reply:
Thanks, that explains a lot.